services/datamanager/builtin/builtin_test.go

```diff
@@ -332,6 +339,105 @@ func TestRecoversAfterKilled(t *testing.T) {
 	test.That(t, len(mockService.getUploadedFiles()), test.ShouldEqual, 1+numArbitraryFilesToSync)
 }
 
+// TODO(DATA-341): Handle partial downloads in order to resume deployment.
+// TODO(DATA-344): Compare checksum of downloaded model to blob to determine whether to redeploy.
+// Want to make sure user can deploy to a custom location
+// Validates that models can be deployed onto a robot.
+func TestModelDeploy(t *testing.T) {
+	deployedZipFileName := "model.zip"
+	originalFileName := "model.txt"
+	otherOriginalFileName := "README.md"
+
+	zipper, err := os.Create(deployedZipFileName)
+	test.That(t, err, test.ShouldBeNil)
+
+	// Add content to zip file.
+	b0 := []byte("text representing model.txt internals.")
+	r0 := bytes.NewReader(b0)
+	b1 := []byte("text representing README.md internals.")
+	r1 := bytes.NewReader(b1)
+	zipWriter := zip.NewWriter(zipper)
+
+	zipFile1, err := zipWriter.Create(originalFileName)
+	test.That(t, err, test.ShouldBeNil)
+	_, err = io.Copy(zipFile1, r0)
+	test.That(t, err, test.ShouldBeNil)
+
+	zipFile2, err := zipWriter.Create(otherOriginalFileName)
+	test.That(t, err, test.ShouldBeNil)
+	_, err = io.Copy(zipFile2, r1)
+	test.That(t, err, test.ShouldBeNil)
+
+	// Now that we are done editing the zip file we close.
+	zipWriter.Close()
```

Comment:
ok. Now we have
```
	zipper, err := os.Create(deployedZipFileName)
	test.That(t, err, test.ShouldBeNil)
	defer zipper.Close()
```
However I cannot do:
```
	zipWriter := zip.NewWriter(zipper)
	defer zipWriter.Close()
```
When I `defer zipWriter.Close()`, I error at `zip.OpenReader(fileNamePath)` inside `model/unzipSource()`.

After digging around I realized that if we have `defer zipWriter.Close()` then the reason why we error  inside `model/unzipSource()` is because the zipfiles themselves do not contain any data - they are 0kB. This means that there is something wrong with `func (m *mockClient) Do`. However the code in that function is correct..? inside that function we have `buf, err := os.ReadFile(m.zipFileName)` but we cannot read a file if it has not been closed.

Reply:
I believe the fix may be:
```
	// Add content to zip file.
	b0 := []byte("text representing model.txt internals.")
	r0 := bytes.NewReader(b0)
	b1 := []byte("text representing README.md internals.")
	r1 := bytes.NewReader(b1)

        // Create zip file.
	deployedZipFile, err := os.Create(deployedZipFileName)
	test.That(t, err, test.ShouldBeNil) 
	zipWriter := zip.NewWriter(zipper)

	defer deployedZipFile.Close()
	defer zipWriter.Close()
```
Defers happen in a LIFO (last in first out) order. 

_but_ after reading around, https://www.joeshaw.org/dont-defer-close-on-writable-files/ indicates that you should explicitly close files rather than deferring, and just make sure to close on any errors.

Reply:
hmm does not work :/.. still throws the same error. I will stick with the explicit close.

Reply:
The reason that deferring `zipWriter.Close()` caused the issue with 0kb data is that Close finishes writing the zip file by writing the central directory, and does not close the actual underlying writer (which is why you still call `deployedZipFile.Close()`). Based on that, the current explicit close after writing is correct.

Reply:
kk